### PR TITLE
Organize as python package

### PR DESCRIPTION
**Pull request by xylar**
@tangq and @golaz, here would be my suggestion for how to reorganize this into a python package (which could also eventually be a conda package).

### CODE REVIEW

**Comment by xylar** [comment]
I have verified that I can run the following locally.  First, install Mambaforge for Linux x86_64 https://github.com/conda-forge/miniforge#mambaforge.  You probably don't want to let it alter your `.bashrc.`.  Assuming it's in `~/mambaforge`:
```
source ~/mambaforge/etc/profile.d/conda.sh
source ~/mambaforge/etc/profile.d/mamba.sh
mamba create -y -n zppy_dev python=3.11 zppy=2.2.0
mamba activate zppy_dev
python -m pip install .
```
After that, I can see that the file are in the conda environment:
```
$ cd ~/mambaforge/envs/zppy_dev/lib/python3.11/site-packages/zppy_e3sm_chem_diags
$ ls
e3sm_chem_diags.py  __init__.py  __pycache__  templates
$ ls templates
chem_summary_table.bash            e3sm_pres_lat_plots.bash
e3sm_chem_cmip_comparison.bash     __init__.py
e3sm_chem_diags.bash               o3_hole_diags.bash
e3sm_chem_diags_ts.bash            __pycache__
e3sm_chem_index.bash               STE_flux_diags.bash
e3sm_chem_noaa_co_comparison.bash  surf_o3_diags.bash
e3sm_lat_lon_plots.bash            temperature_eq_plot.bash
e3sm_nox_emis_plots.bash           TOZ_eq_plot.bash
```

**Comment by xylar** [comment]
The next step would be to modify https://github.com/E3SM-Project/zppy/pull/402 to work with this.  My suggestion would be that, rather than giving a path like the following for the plugin:
```
/home/xylar/mambaforge/envs/zppy_dev/lib/python3.11/site-packages/zppy_e3sm_chem_diags/e3sm_chem_diags.py
```
you would simply give the plugin name (`plugin = zppy_e3sm_chem_diags`) and `zppy` would find it using `importlib.resources`.

**Comment by tangq** [comment]
@xylar , thanks for making ChemDyg a python package and suggestions for eventually including it in e3sm_unified. None of our chemistry team member is familiar with this process, but I am sure that it's in good hands (you, @golaz , and @chengzhuzhang). Please feel free to add necessary changes.

**Comment by xylar** [comment]
@tangq, that's precisely my concern. I don't have the time, and I suspect neither do @chengzhuzhang and @golaz to become responsible for this package. Ut is my feeling that someone from the chemistry team needs to put the time in to develop this expertise. This isn't a service the infrastructure team can provide.

**Comment by tangq** [comment]
@xylar , I agree that it would be better that the chemistry team have this expertise. Let's talk about this at our meeting next week.

**Comment by tangq** [comment]
@xylar , @chengzhuzhang , @golaz, to follow up on what we discussed at the [infrastructure meeting](https://acme-climate.atlassian.net/wiki/spaces/IPD/pages/3704488004/2023-03-09+Analysis+Tools+Meeting+notes), we need to complete these steps (Please make corrections if needed) for including ChemDyg in e3sm_unified:

1. Make `ChemDyg` a python package. (Done by @xylar as described above. Thanks!)
2. Submit the `ChemDyg` package to the `conda-forge` channel. (To be done)
3. Modify [zppy plugin](https://github.com/E3SM-Project/zppy/pull/402) to work with this and use `importlib.resources` to find the path of plugins. (To be done)

In this `ChemDyg` space, I think we should focus on 2 and leave the discussions on 3 in the `zppy` space.

I found this `conda-forge` [instruction page](https://conda-forge.org/docs/maintainer/adding_pkgs.html#the-staging-process) and it seems that the next step for 2 is to create the recipe file (meta.yaml) using `graskull`. I will try that later today.

**Comment by xylar** [comment]
@tangq, I'm happy to create the recipe. Grayskull won't help in this case. It's very handy for packages already available on PyPI, the server for the pip installation tool. But we won't use PyPI so the meta.yaml needs to be created from scratch.

You do need to release the package, since conda-forge recipes need to refer to a release version of the python package.

**Comment by tangq** [comment]
@xylar , if I create a release following these [instructions](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository), it will be based on the current main without the commits you made in this PR to make it a python package. Shall we merge this PR before the release, which will be referred to by the recipe?

**Comment by xylar** [comment]
@tangq, yes, those are the right instructions, but we would want to merge this PR first.

Let me rebase to pick up your recent changes.  Then, you and @hsiangheleellnl can review.

**Comment by xylar** [comment]
Okay @tangq and @hsiangheleellnl, this is ready for review.  I will make a few notes for your to respond to.

File: .gitignore
Comment (xylar) [inline .gitignore:1]: This is a standard file to make sure you only commit the things you want.  In particular, it makes sure to ignore the build artifacts when you build and install the python package into a conda environment.

File: MANIFEST.in
Comment (xylar) [inline MANIFEST.in]: This is how we make sure the the `.bash` files get included in the python package (by default, only `.py` files are included). If you add more non-python files to the package that `zppy` needs to be able to fine, you should add them here, too.

File: setup.cfg
Comment (xylar) [inline setup.cfg:1]: This file is one of the vital ingredients in making an installable python package.  Here, you say what version the package will have.  before you release the package, you need to update the version here.

File: setup.cfg
Comment (xylar) [inline setup.cfg]: This is a very important point for us to decide on now before there is a release: what should the name of the python package be?  This was my suggestion but it could also just be `e3sm-chem-diags` or `e3sm_chem_diags`.  Please let me know what you prefer.

File: setup.cfg
Comment (xylar) [inline setup.cfg]: The version could also be 0.1.0, indicating that it is not a fully mature package yet.  I strongly encourage using this version numbering scheme:
https://semver.org/
If you do, the first number of 3 (the major version) is updated when there are changes that break backward compatibility with previous versions.  The second number (the minor version) is for changes that don't break backwards compatibility, so if you update from 1.0.0 to 1.1.0, we would expect that `zppy` still works as before.  This means that a given version of zppy should be able to depend on `zppy-e3sm-chem-diags >=1.0.0,<2.0.0` and know that nothing will be broken.  The third number (the micro version) is used for bug fixes only, so 1.0.1 would have a bug fix on 1.0.0.

File: setup.cfg
Comment (xylar) [inline setup.cfg]: Please suggest changes as you see fit.  I was following the repo as best I could. 

@golaz and @forsyth2 are both authors of zppy, so it is fine for the authors and emails to be a comma-separated list.  See:
https://github.com/E3SM-Project/zppy/blob/main/setup.py#L27-L28
but note that the format for `setup.py` and `setup.cfg` are different.

Comment (xylar) [reply setup.cfg]: I wonder if we might want to add `E3SM` to the description:
```suggestion
author = Hsiang-He Lee
author_email = lee1061@llnl.gov
description = E3SM Chemistry Diagnostics Package
```

File: setup.cfg
Comment (xylar) [inline setup.cfg:36]: We will need to increase this constraint once there is a release of `zppy` that supports plugins.

We should add more dependencies here.  This is one of the reasons I would like to pull out the python files, rather than embedding them in the bash files.  It would be much easier to determine what dependencies are required if they were always at the top of the python files.

File: setup.py
Comment (xylar) [inline setup.py:1]: This is another file that makes the python package installable.

Comment (xylar) [reply setup.py:1]: The `zppy_e3sm_chem_diags/__init__.py` is the final file that makes this a python package.  It is currently empty but it is required.  You can add things to it later.  We can talk about why you might want to do that.  (It can be used to simplify imports from this package, e.g. within `zppy`.)

File: setup.cfg
Comment (xylar) [reply setup.cfg]: Currently, there is a directory `zppy_e3sm_chem_diags` where I have moved a lot of files.  This is the python package itself.  Any files outside of this directory are in the repository but not in the python package.  So keep that in mind.  None of them will be available to `zppy` or to users that have installed the conda package.

Also, I will note that we will rename `zppy_e3sm_chem_diags` to correspond to the name you choose here.  Hyphens are not allowed in the directory name (or any directories or python files in a python package, since they get confused with minus signs in python imports).  So it is common for the package name to have hyphens but the directory to have underscores like in my example.  But both could have underscores if you prefer.

The name of the python package should be all lowercase as I have here (so shouldn't be ChemDyg, for example).

File: setup.py
Comment (xylar) [reply setup.py:1]: Each directory (e.g. `templates`) within the package also needs a `__init__.py` so it is included in the package.  Keep that in mind if you add new directories within the package.

Also, as I noted above, all files outside of `zppy_e3sm_chem_diags` are not considered to be part of the python package.

**Comment by xylar** [comment]
@tangq and @hsiangheleellnl, one more thing.  I wouldn't do a release or create a conda package just yet.  I think it makes sense to first get a release of `zppy` that fully supports plugins.  The release of this package should depend on a version of `zppy` that can make use of it.

In the meantime, you can install zppy following the instructions for developers:
https://e3sm-project.github.io/zppy/_build/html/main/getting_started.html#dev-env
Then, you can go to your local clone of this repository and run:
```
python -m pip install -e .
```
This will install this python package (`zppy-e3sm-chem-diags` or `e3sm-chem-diags` or whatever we decide to call it) into the zppy conda environment in "editable" mode.  That means that zppy will see the version of the code in your local clone, so if you make changes you don't need to rerun the install command for zppy to make use of those changes.

This is the typical way we would develop 2 python packages at the same time.

File: setup.cfg
Comment (xylar) [reply setup.cfg:36]: I believe these are all your dependencies.  Please add version constraints if you know them.  This is especially important if you don't think your code is compatible with the latest versions of these packages (or if you suspect it won't be compatible with a future release).  Sometimes you will only learn this when it's too late (and it's pretty inconvenient).  So we can update these as we need to.
```suggestion
   cartopy
   matplotlib
   netcdf4
   numpy
   pandas
   scipy
   xarray
   zppy >=2.2.0
```

Comment (xylar) [reply setup.cfg:36]: I went ahead and committed these changes.

**Comment by tangq** [comment]
@xylar , thanks a lot. I will go through your notes tomorrow.

> Okay @tangq and @hsiangheleellnl, this is ready for review. I will make a few notes for your to respond to.

**Comment by chengzhuzhang** [comment]
@xylar Thank you so much for your comments and note here. I found them very helpful. And I will plan to summarize the key steps into a confluence page, so other developers such as @susburrows who are interested in contributing new packages can follow.

File: setup.cfg
Comment (chengzhuzhang) [reply setup.cfg]: @xylar the names proposed here are all good. If there is no strong preference,  I think in general it would be easier to avoid using dash `-` in the package name, but to use underscore instead, so the module name and the package name are consistent.

Comment (xylar) [reply setup.cfg]: I agree that underscores do tend to avoid a certain level of confusion.  For some reason, the community seems to prefer dashes over underscores.

File: setup.cfg
Comment (hsiangheleellnl) [reply setup.cfg]: I am okay with the name "zppy_e3sm_chem_diags".

File: setup.cfg
Comment (hsiangheleellnl) [inline setup.cfg]:
```suggestion
version = 0.1.0
```

**Review by hsiangheleellnl — APPROVED** [review]
@xylar Thank you for the help and the detailed instruction.

File: setup.cfg
Comment (hsiangheleellnl) [inline setup.cfg]:
```suggestion
name = zppy_e3sm_chem_diags
```

**Review by hsiangheleellnl — APPROVED** [review]

File: .gitignore
Comment (tangq) [reply .gitignore:1]: This is very useful as when I pushed the initial version to github, I had to manually remove files.

File: MANIFEST.in
Comment (tangq) [reply MANIFEST.in]: @xylar , do you add this `MANIFEST.in` file manually or by some tool (e.g., `mambaforge`)?

Sorry for many elementary questions as I went through this for the first time. And I do feel knowing more details will help maintain ChemDyg in the long-term.

File: setup.cfg
Comment (tangq) [reply setup.cfg:1]: I will keep a note on this. Thanks!

File: MANIFEST.in
Comment (xylar) [reply MANIFEST.in]: I added this manually myself and you need to update it manually yourselves as you add more non-python files that need to be included in the package.

Comment (xylar) [reply MANIFEST.in]: And these are very reasonable questions.

**Comment by xylar** [comment]
Okay, I've made your 2 changes @hsiangheleellnl.  I think that answers my questions for now.  Please merge when you're happy.

File: setup.cfg
Comment (tangq) [reply setup.cfg]: Since python only allows underscore in the directory name, I'd prefer only use underscore in the names of directory and python package for consistency.

The name we came up for the E3SM chemistry diagnostic package is `ChemDyg` as the github repo name. I think we would like to use `ChemDyg` for the python package name as well.

With the all lowercase requirement, I am leaning towards `chemdyg` as the python package name, which will need these changes:

- Rename the directory `zppy_e3sm_chem_diags` to a new name. Any suggestions for the new directory name? When cloning from github, we will have the `ChemDyg` directory, under which is the python package directory. Is there a standard naming convention? If not, what about `chemdyg_package` (I feel it's a bit weird if we have `chemdyg` under `ChemDyg`)?
- Change to `name = chemdyg` in this setup.cfg file

Welcome to suggestions.

File: MANIFEST.in
Comment (hsiangheleellnl) [reply MANIFEST.in]: I see. I will change them soon. @tangq

Comment (hsiangheleellnl) [reply MANIFEST.in]: @xylar I would like to double check here. Do I need to list all the .bash files one by one in this file? not like you did using *.bash?

File: setup.cfg
Comment (tangq) [reply setup.cfg]: Sorry that I had a few meetings this morning and couldn't add my two cents here earlier. Please see my suggestions on the python package name and directory name above.

Comment (hsiangheleellnl) [reply setup.cfg]: If so, I prefer to rename the directory to "chemdyg_package". 
How about the file name: e3sm_chem_diags.py? Should it change to "chemdyg.py" or "chemdyg_package.py" as well?

Comment (tangq) [reply setup.cfg]: I like `chemdyg.py`. Certainly not `chemdyg_package.py`.

File: setup.cfg
Comment (tangq) [inline setup.cfg]: I'd suggest revising to 
```suggestion
author = Hsiang-He Lee, Qi Tang, Michael J. Prather, Jean-Christophe Golaz
```
for their contributions of the science and metrics design, original code and examples, and refactor the link with zppy.

File: setup.py
Comment (tangq) [reply setup.py:1]: Got it. `__init__.py` is needed for each directory in the package.

I assume the information (except format) should be consistent between `setup.cfg` and `setup.py`, which is handled manually.

**Comment by tangq** [comment]
This is very good tip - Thank you, @xylar .

> @tangq and @hsiangheleellnl, one more thing. I wouldn't do a release or create a conda package just yet. I think it makes sense to first get a release of `zppy` that fully supports plugins. The release of this package should depend on a version of `zppy` that can make use of it.
> 
> In the meantime, you can install zppy following the instructions for developers: https://e3sm-project.github.io/zppy/_build/html/main/getting_started.html#dev-env Then, you can go to your local clone of this repository and run:
> 
> ```
> python -m pip install -e .
> ```
> 
> This will install this python package (`zppy-e3sm-chem-diags` or `e3sm-chem-diags` or whatever we decide to call it) into the zppy conda environment in "editable" mode. That means that zppy will see the version of the code in your local clone, so if you make changes you don't need to rerun the install command for zppy to make use of those changes.
> 
> This is the typical way we would develop 2 python packages at the same time.

File: setup.cfg
Comment (xylar) [reply setup.cfg]: Nearly all python package name the directory the same as the package name.  So it would be appropriate to name the package `chemdyg` and the directory `chemdyg` but it would be very confusing to a python programmer if you called the directory `chemdyg_package`, since that's quite non-standard.  Some people call the directory something like `src` or `lib` but I think that's pretty poor form and doesn't make things less confusing.  I would strongly recommend just `chemdyg` for both.

I will day that `dyg` doesn't mean anything to me so it would be helpful to have some explanation of what that is intended to mean.  Somehow short for "diagnostics"?

Comment (xylar) [reply setup.cfg]: I do think `e3sm_chem_diags.py` be named either `chemdyg.py`.  That somewhat depends on what zppy will require when it is looking for plugins.  Another possibility would be to put it in `__init__.py`.  That would be intuitive for a python package programmer but won't make much sense to people new to python packaging.  So I won't recommend that approach.

File: setup.py
Comment (xylar) [reply setup.py:1]: Yes, the same information is needed in either `setup.py` or `setup.cfg`.  Either place is fine.  This package is pretty simple so I think `setup.cfg` will be easier.

File: MANIFEST.in
Comment (xylar) [reply MANIFEST.in]: No, you do not need to list all the files.  The `*.bash` is sufficient.  I believe this covers everything right now.  It's just that we would need to update it if you add new types of files or new bash files in a different location or something like that.

**Comment by xylar** [comment]
@tangq and @hsiangheleellnl, are we agreed that the package and directory are to be named `chemdyg`?  @tangq commented that this seems redundant (`chemdyg` inside of `ChemDyg`) but this is *very* common for python packages and just something you have to get used to.

Are we also agreed that (at least for now) `e3sm_chem_diags.py` should become `chemdyg.py`?

**Comment by xylar** [comment]
I went ahead with the change for now.  I can always undo or fix the commit based on your feedback but it is late here so I want to give you the chance to decide today if you want to.

**Review by hsiangheleellnl — APPROVED** [review]

File: setup.cfg
Comment (tangq) [reply setup.cfg]: @xylar , we should follow the common practice to use the same name `chemdyg` for the package and the directory. I suggested the "_package" suffix because git clone will create a `chemdyg` directory if the users do not give it a different name. I am fine with the path like `chemdyg/chemdyg`.

Yes, `dyg` stands for diagnostics. We can describe it in the README.md file.

It seems that we all think it is a good idea to rename the python file to `chemdyg.py`.

File: setup.py
Comment (tangq) [reply setup.py:1]: Okay, let's just use `setup.cfg`.

**Comment by tangq** [comment]
Yes, these changes are all good to me. Thanks.

> @tangq and @hsiangheleellnl, are we agreed that the package and directory are to be named `chemdyg`? @tangq commented that this seems redundant (`chemdyg` inside of `ChemDyg`) but this is _very_ common for python packages and just something you have to get used to.
> 
> Are we also agreed that (at least for now) `e3sm_chem_diags.py` should become `chemdyg.py`?

**Comment by tangq** [comment]
As @xylar recommended above we would wait until `zppy` supports this plugin to create a release or a conda package for `chemdyg`. But is this PR ready to be merged now? If so, I will click on the `Squash and merge` button.

**Review by tangq — APPROVED** [review]

File: setup.py
Comment (mahf708) [reply setup.py:1]: Just a side comment: for future plugins/packages, it may be worthwhile simplifying this even further by using pyproject.toml instead of setup.cfg (and thus likely eliminating the need for the additional setup.py file).

Comment (xylar) [reply setup.py:1]: I don't feel very comfortable with `pyproject.toml` files yet, which is why I didn't go for that solution.  I have more trouble maintaining packages on conda-forge that use that format than the older approaches (though that has more to do with grayskull not supporting the `pyproject.toml` format).  Still, if someone wants to give that a go, that's fine.

Comment (mahf708) [reply setup.py:1]: Oh, yeah I remember... many feedstocks used to have `rm -rf pyproject.toml` before `{ PYTHON } -m pip install . -vvv` because pyproject.toml used to cause a lot of issues. Anyway, using static package setups (setup.cfg or pyproject.toml) is much better than the dynamic setup.py and so we are good here :D

**Pull request merged**
